nixos/modules/services/hardware/pid-fan-controller.nix

```
@@ -0,0 +1,189 @@
+        '';
+      };
+      fans = lib.mkOption {
+        type = lib.types.listOf (lib.types.submodule fan);
+        description = "List of fans to be controlled.";
+        example = ''
+          [
+            {
+              wildcardPath = "/sys/devices/platform/nct6775.2592/hwmon/hwmon*/pwm1";
+              minPwm = 60;
+              maxPwm = 255;
+              heatPressureSrcs = [
+                "cpu"
+                "gpu"
+              ];
+            }
+          ];
+        '';
+      };
+    };
+  };
+  config = lib.mkIf cfg.enable {
+    #map camel cased attrs into snake case for config
+    environment.etc."pid-fan-settings.json".text = builtins.toJSON {
+      interval = cfg.settings.interval;
+      heat_srcs = map (heatSrc: {
+        name = heatSrc.name;
+        wildcard_path = heatSrc.wildcardPath;
+        PID_params = {
+          set_point = heatSrc.pidParams.setPoint;
+          P = heatSrc.pidParams.P;
+          I = heatSrc.pidParams.I;
+          D = heatSrc.pidParams.D;
+        };
+      }) cfg.settings.heatSources;
+      fans = map (fan: {
+        wildcard_path = fan.wildcardPath;
+        min_pwm = fan.minPwm;
+        max_pwm = fan.maxPwm;
+        cutoff = fan.cutoff;
+        heat_pressure_srcs = fan.heatPressureSrcs;
+      }) cfg.settings.fans;
+    };
+
+    systemd.services.pid-fan-controller = {
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = [ (lib.getExe cfg.package) ];
+        ExecStopPost = [ "${lib.getExe cfg.package} disable" ];
+        Restart = "always";
+        #This service needs to run as root to write to /sys.
+        #therefore it should operate with the least amount of privileges needed
+        ProtectHome = "yes";
+        #strict is not possible as it needs /sys
+        ProtectSystem = "full";
+        ProtectProc = "invisible";
+        PrivateNetwork = "yes";
+        NoNewPrivileges = "yes";
+        MemoryDenyWriteExecute = "yes";
+        RestrictNamespaces = "~user pid net uts mnt";
+        ProtectKernelModules = "yes";
+        RestrictRealtime = "yes";
+        SystemCallFilter = "@system-service";
+        CapabilityBoundingSet = "~CAP_KILL CAP_WAKE_ALARM CAP_IPC_LOC CAP_BPF CAP_LINUX_IMMUTABLE CAP_BLOCK_SUSPEND CAP_MKNOD";
+      };
+      # restart unit if config changed
+      restartTriggers = [ config.environment.etc."pid-fan-settings.json".text ];
+    };
+    #sleep hook to restart the service as it breaks otherwise
+    systemd.services.pid-fan-controller-sleep = {
+      before = [ "sleep.target" ];
+      wantedBy = [ "sleep.target" ];
+      unitConfig = {
+        StopWhenUnneeded = "yes";
+      };
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        ExecStart = [ "${config.systemd.package}/bin/systemctl stop pid-fan-controller.service" ];
+        ExecStop = [ "${config.systemd.package}/bin/systemctl restart pid-fan-controller.service" ];
```

```suggestion
        ExecStart = [ "systemctl stop pid-fan-controller.service" ];
        ExecStop = [ "systemctl restart pid-fan-controller.service" ];
```
systemctl should already be in PATH of the unit by default